tempesta_fw/t/unit/tfw_bomber.c

```diff
@@ -0,0 +1,496 @@
+module_param(tfw_server, charp, 0);
+
+MODULE_PARM_DESC(tfw_server, "Server host address and optional port nunber");
+MODULE_LICENSE("GPL");
+
+#define TFW_BOMBER_WAIT_INTVL		(2)		/* in seconds */
+#define TFW_BOMBER_WAIT_MAX		(1 * 60)	/* in seconds */
+
+/* Flags for tfw_bomber_desc_t.flags */
+#define TFW_BOMBER_CONNECT_STARTED	(0x0001)
+#define TFW_BOMBER_CONNECT_ESTABLISHED	(0x0002)
+#define TFW_BOMBER_CONNECT_CLOSED	(0x0004)
+#define TFW_BOMBER_CONNECT_ERROR	(0x0100)
+
+typedef struct tfw_bomber_desc {
+	SsProto		proto;
+	struct sock	*sk;
+	uint32_t	flags;
+} tfw_bomber_desc_t;
+
+/*
+ * There's a descriptor for each connection that keeps the connection's
+ * state and status. All descriptors are kept in a static two-dimensional
+ * array. SsProto.type field is used here to store the index into that
+ * array that can be passed around between callbacks.
+ */
+static tfw_bomber_desc_t **tfw_bomber_desc;
+
+static struct task_struct **tfw_bomber_connect_task;
+static struct task_struct *tfw_bomber_finish_task;
+
+DECLARE_WAIT_QUEUE_HEAD(tfw_bomber_connect_wq);
+DECLARE_WAIT_QUEUE_HEAD(tfw_bomber_finish_wq);
+
+static atomic_t tfw_bomber_nthread;
+static atomic_t tfw_bomber_connect_nattempt;  /* Successful attempts */
+static atomic_t tfw_bomber_connect_ncomplete; /* Connections established */
+static atomic_t tfw_bomber_connect_nerror;    /* Number of errors */
+static atomic_t tfw_bomber_request_nsend;     /* Number of requests */
+
+static TfwAddr tfw_bomber_server_address;
+static SsHooks tfw_bomber_hooks;
+
+static int
+tfw_bomber_connect(int descidx)
+{
+	int ret;
+	struct sock *sk;
+	tfw_bomber_desc_t *desc = *(tfw_bomber_desc + descidx / tfw_connects)
+						    + descidx % tfw_connects;
+
+	ret = ss_sock_create(tfw_bomber_server_address.sa.sa_family,
+			     SOCK_STREAM, IPPROTO_TCP, &sk);
+	if (ret) {
+		SS_DBG("Unable to create kernel socket (%d)\n", ret);
+		desc->flags |= TFW_BOMBER_CONNECT_ERROR;
+		atomic_inc(&tfw_bomber_connect_nerror);
+		return ret;
+	}
+	ss_proto_init(&desc->proto, &tfw_bomber_hooks, descidx);
+	sk->sk_user_data = &desc->proto;
+	ss_set_callbacks(sk);
+	local_bh_disable();
+	ret = ss_connect(sk, &tfw_bomber_server_address.sa,
+			 tfw_addr_sa_len(&tfw_bomber_server_address), 0);
+	if (ret) {
+		SS_DBG("Connect error on server socket sk %p (%d)\n", sk, ret);
+		sk->sk_user_data = NULL;
+		ss_close(sk);
+		local_bh_enable();
+		desc->flags |= TFW_BOMBER_CONNECT_ERROR;
+		atomic_inc(&tfw_bomber_connect_nerror);
+		return ret;
+	}
+
+	local_bh_enable();
+	desc->sk = sk;
+	desc->flags |= TFW_BOMBER_CONNECT_STARTED;
+	atomic_inc(&tfw_bomber_connect_nattempt);
+
+	return 0;
+}
+
+static void
+msg_send(tfw_bomber_desc_t *desc)
+{
+	struct sock *sk;
+	char *str;
+	int len, ret;
+	TfwStr msg;
+	TfwHttpMsg *req;
+	TfwMsgIter it;
+
+	sk = desc->sk;
+	BUG_ON(!sk);
+
+	atomic_inc(&tfw_bomber_request_nsend);
+
+	len = 1 * 1024 * 1024;
+	str = vmalloc(len);
```

Comment:
This is the killer for performance. Just create an array of ```void *``` with size of ```tfw_threads```, and preallocate memory for each member of the array. After than each thread can use the memory regions addressing them by their identifiers.